src/WebJobs.Script/ScriptHostBuilderExtensions.cs

```diff
@@ -157,12 +159,16 @@ public static IHostBuilder AddScriptHostCore(this IHostBuilder builder, ScriptAp
 
         public static void AddCommonServices(IServiceCollection services)
         {
+            // The scope for these services is beyond a single host instance.
+            // They are not recreated for each new host instance, so you have
+            // to be careful with caching, etc. E.g. these services will get
+            // initially created in placeholder mode and live on through the
+            // specialized app.
             services.AddSingleton<IHostIdProvider, ScriptHostIdProvider>();
             services.TryAddSingleton<IScriptEventManager, ScriptEventManager>();
             services.TryAddSingleton<IDebugManager, DebugManager>();
             services.TryAddSingleton<IDebugStateProvider, DebugStateProvider>();
             services.TryAddSingleton<IEnvironment>(SystemEnvironment.Instance);
-            services.AddSingleton<IProxyMetadataManager, ProxyMetadataManager>();
```

Comment:
This is the fix. Previously this was a "common" service and it read and cached proxy metadata while in placeholder mode. The service was never created after specialization so the cached data was returned.

Reply:
I do notice a small issue with this change as coded - WebFunctionsManager which is a management level service relies on this proxy service (see [here](IWebFunctionsManager)). The fix is to change that location to use a static method just like we do for FunctionMetadataManager.ReadFunctionsMetadata in that same function, thus severing the dependency.

Reply:
I'll let @fabiocav weigh in on this as I can't recall why this was moved to the CommonServices. But we have a pattern elsewhere (like secrets) where we reset the cached values when specialization occurs: https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script.WebHost/Security/KeyManagement/DefaultSecretManagerProvider.cs#L40-L41. That may be a more targeted fix.

I do agree that it's troublesome how easy it is to get these scopes wrong...

Reply:
Reacting to the options change is a viable approach. I agree that scope validation on these would be helpful. I believe this was moved here when there was an iteration of the function metadata API that exposed proxies.